src/python/grpcio_csm_observability/grpc_csm_observability/_csm_observability_plugin.py

```diff
@@ -0,0 +1,373 @@
+        for example, if the channel target string uses IP addresses directly).
+        Return True means the original target string will be used, False means target string
+        will be replaced with "other".
+          generic_method_attribute_filter: Once provided, this will be called with a generic
+        method type to decide whether to record the method name or to replace it with
+        "other". Note that pre-registered methods will always be recorded no matter what
+        this function returns.
+        Return True means the original method name will be used, False means method name will
+        be replaced with "other".
+        """
+        self.plugin_options = list(plugin_options) + [
+            CsmOpenTelemetryPluginOption()
+        ]
+        self.meter_provider = meter_provider
+        if target_attribute_filter:
+            self.target_attribute_filter = target_attribute_filter
+        else:
+            self.target_attribute_filter = lambda target: True
+        if generic_method_attribute_filter:
+            self.generic_method_attribute_filter = (
+                generic_method_attribute_filter
+            )
+        else:
+            self.generic_method_attribute_filter = lambda method: False
+        self._plugins = [
+            _open_telemetry_observability._OpenTelemetryPlugin(self)
+        ]
+
+    def _get_enabled_optional_labels(self) -> List[OptionalLabelType]:
+        return [OptionalLabelType.XDS_SERVICE_LABELS]
+
+
+def get_value_from_struct(key: str, struct: struct_pb2.Struct) -> str:
+    value = struct.fields.get(key)
+    if not value:
+        return UNKNOWN_VALUE
+    return value.string_value
+
+
+def get_str_value_from_resource(
+    attribute: Union[ResourceAttributes, str], resource: Resource
+) -> str:
+    value = resource.attributes.get(attribute, UNKNOWN_VALUE)
+    return str(value)
+
+
+def get_resource_type(gcp_resource: Resource) -> str:
+    # Convert resource type from GoogleCloudResourceDetector to the value we used for
+    # metadata exchange.
+    # Reference: http://shortn/_eY0hMnWTVd
+    gcp_resource_type = get_str_value_from_resource(
+        "gcp.resource_type", gcp_resource
+    )
+    if gcp_resource_type == "gke_container":
+        return TYPE_GKE
+    elif gcp_resource_type == "gce_instance":
+        return TYPE_GCE
+    else:
+        return gcp_resource_type
+
+
+# Returns the mesh ID by reading and parsing the bootstrap file. Returns "unknown"
+# if for some reason, mesh ID could not be figured out.
+def get_mesh_id() -> str:
+    config_contents = get_bootstrap_config_contents()
+
+    try:
+        config_json = json.loads(config_contents)
+        # The expected format of the Node ID is -
+        # projects/[GCP Project number]/networks/mesh:[Mesh ID]/nodes/[UUID]
+        node_id_parts = config_json.get("node", {}).get("id", "").split("/")
+        if len(node_id_parts) == 6 and "mesh:" in node_id_parts[3]:
```

Comment:
I synced with other language teams and we were all doing different granularities of assertions here. C++ Java and Go all split into parts based off /, but then had different requirements of what would fail and become unknown.